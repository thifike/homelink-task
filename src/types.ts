/**
 * Enums
 */
export enum DeviceType {
  EnvironmentalSensor = "ES",
  HeatAlarm = "HA",
  SmokeAlarm = "SA",
}

export enum OperationalStatus {
  Faulty = "F",
  Operational = "O",
}

export enum AlarmStatus {
  Normal = "N",
  Alarming = "A",
}

/**
 * Types
 */

export interface ThermalDataPoint {
  id: string;
  dataPointType: "T";
  temperature: number;
  alarmStatus: AlarmStatus;
}

export interface EnvironmentalDataPoint {
  id: string;
  dataPointType: "E";
  temperature: number;
  humidity: number;
}

export interface SmokeDataPoint {
  id: string;
  dataPointType: "S";
  concentration: number;
  alarmStatus: AlarmStatus;
}

export type DataPoint = ThermalDataPoint | EnvironmentalDataPoint | SmokeDataPoint;

export interface ThermalDeviceConfiguration {
  deviceType: DeviceType.HeatAlarm;
  temperatureLowThreshold: number;
  temperatureHighThreshold: number;
}

export interface EnvironmentalDeviceConfiguration {
  deviceType: DeviceType.EnvironmentalSensor;
}

export interface SmokeDeviceConfiguration {
  deviceType: DeviceType.SmokeAlarm;
  concentrationHighThreshold: number;
}

export type DeviceConfiguration =
  | ThermalDeviceConfiguration
  | EnvironmentalDeviceConfiguration
  | SmokeDeviceConfiguration;

export interface DeviceCreationParams {
  /**
   * @minLength 3 Series information must not be shorter than 3 characters!
   */
  series: string;
  /**
   * @minLength 3 Model name must not be shorter than 3 characters!
   */
  modelName: string;
  deviceConfiguration:
    | ThermalDeviceConfiguration
    | EnvironmentalDeviceConfiguration
    | SmokeDeviceConfiguration;
}

export interface Device extends DeviceCreationParams {
  id: string;
}

export interface ValidateErrorJSON {
  message: "Validation failed";
  details: { [name: string]: unknown };
}
