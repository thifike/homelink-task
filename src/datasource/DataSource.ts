export type IdGeneratorFunction<TId> = () => TId;

export interface IdType<T> {
  id: T;
}

export abstract class DataSource<TId, T = { id: TId }> {
  public abstract create(input: T): Promise<T>;
  public abstract read(id: TId): Promise<T>;
  public abstract list(predicate?: (item: T, index: number, array: T[]) => boolean): Promise<T[]>;
  public abstract update(item: T): Promise<T>;
  public abstract delete(id: TId): Promise<boolean>;
}
