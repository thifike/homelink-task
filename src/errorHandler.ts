import { NextFunction, Request, Response } from "express";
import { ValidateError } from "tsoa";

export function errorHandler(err: unknown, req: Request, res: Response, next: NextFunction): void {
  if (err instanceof ValidateError) {
    console.warn(`Caught Validation Error for ${req.path}:`, err.fields);
    res.status(422).json({
      message: "Validation Failed",
      details: err?.fields,
    });
    return;
  }
  if (err instanceof Error) {
    res.status(500).json({
      message: err.message,
      stack: err.stack,
    });
    return;
  }

  next();
}
