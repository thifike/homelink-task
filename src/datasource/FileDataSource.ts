import { existsSync } from "fs";
import { DataSource, IdGeneratorFunction, IdType } from "./DataSource";
import { readFile, writeFile } from "fs/promises";

export class FileDataSource<TId, T extends IdType<TId>> extends DataSource<TId, T> {
  private filePath: string;
  private idGenerator: IdGeneratorFunction<TId>;

  constructor(options: { filePath: string; idGenerator: IdGeneratorFunction<TId> }) {
    super();

    const { filePath, idGenerator } = options;

    this.filePath = filePath;
    if (typeof idGenerator !== "function") {
      throw new Error("The idGenerator must be a function!");
    }
    this.idGenerator = idGenerator;
  }

  private async readDatabase(): Promise<T[]> {
    if (!existsSync(this.filePath)) {
      await this.writeDatabase([]);
    }

    return JSON.parse(await readFile(this.filePath, { encoding: "utf-8" }));
  }

  private async writeDatabase(data: T[]): Promise<void> {
    await writeFile(this.filePath, JSON.stringify(data));
  }

  public async create(input: T): Promise<T> {
    const id: TId = this.idGenerator();

    const data = await this.readDatabase();

    const item: T = { ...input, id };
    data.push(item);

    await this.writeDatabase(data);

    return item;
  }

  public async read(id: TId): Promise<T> {
    const data = await this.readDatabase();

    const item: T | undefined = data.find((item) => item.id === id);
    if (!item) {
      throw new Error(`Item with id '${id}' not found!`);
    }

    return item;
  }

  public async list(
    predicate?: ((item: T, index: number, array: T[]) => boolean) | undefined
  ): Promise<T[]> {
    const data = await this.readDatabase();

    return data.filter(predicate ?? ((_item) => true));
  }

  public async update(item: T): Promise<T> {
    const data = await this.readDatabase();
    const updateIndex: number = data.findIndex((value) => value.id === item.id);

    if (updateIndex === -1) {
      throw new Error(`Item with id '${item.id}' not found!`);
    }

    data[updateIndex] = item;

    await this.writeDatabase(data);

    return item;
  }

  public async delete(id: TId): Promise<boolean> {
    const data = await this.readDatabase();
    const deleteIndex: number = data.findIndex((value) => value.id === id);

    if (deleteIndex === -1) {
      throw new Error(`Item with id '${id}' not found!`);
    }

    data.splice(deleteIndex, 1);

    await this.writeDatabase(data);

    return true;
  }
}
