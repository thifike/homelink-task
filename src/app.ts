// src/app.ts
import express, { Request as ExRequest, Response as ExResponse, json, urlencoded } from "express";
import swaggerUi from "swagger-ui-express";
import { RegisterRoutes } from "./generated/routes";
import { errorHandler } from "./errorHandler";

export const app = express();

// Use body parser to read sent json payloads
app.use(
  urlencoded({
    extended: true,
  })
);
app.use(json());

// Add Swagger UI
app.use("/docs", swaggerUi.serve, async (_req: ExRequest, res: ExResponse) => {
  res.send(swaggerUi.generateHTML(await import("./generated/swagger.json")));
});

// Register TSOA routes
RegisterRoutes(app);

// 404 handler
app.use((_req, res: ExResponse) => {
  res.status(404).send({
    message: "Not Found",
  });
});

// Global error handler
app.use(errorHandler);
