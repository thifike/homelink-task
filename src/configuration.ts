import { randomUUID } from "crypto";
import { FileDataSource } from "./datasource/FileDataSource";
import { DeviceService } from "./devices/DeviceService";
import { Device } from "./types";

export const DEVICE_SERVICE: DeviceService = new DeviceService({
  deviceDataSource: new FileDataSource<string, Device>({
    filePath: "database/database.json",
    idGenerator: () => randomUUID(),
  }),
});
