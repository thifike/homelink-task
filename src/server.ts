import { app } from "./app";

const port = process.env.PORT || 3000;

app.listen(port, () => console.log(`HomeLink app listening at http://localhost:${port}`));
