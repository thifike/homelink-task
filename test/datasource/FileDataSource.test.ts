const mockReadFile = jest.fn();
const mockWriteFile = jest.fn();
const mockExistsSync = jest.fn();

jest.mock("fs/promises", () => ({
  ...jest.requireActual("fs/promises"),
  readFile: (_path: string, _options: { encoding: string }) => mockReadFile(),
  writeFile: (_path: string, data: string) => mockWriteFile(data),
}));

jest.mock("fs", () => ({
  ...jest.requireActual("fs"),
  existsSync: (_path: string) => mockExistsSync(),
}));

import { readFileSync } from "fs";
import { IdGeneratorFunction } from "../../src/datasource/DataSource";
import { FileDataSource } from "../../src/datasource/FileDataSource";
import { Device, DeviceType } from "../../src/types";

const DEFAULT_DATABASE: string = readFileSync("./test/datasource/default-database.json").toString();

describe("FileDataSource creation tests", () => {
  it("Creation with an id generator function is successful", () => {
    const fds: FileDataSource<string, Device> = new FileDataSource<string, Device>({
      filePath: "somefile.json",
      idGenerator: () => "the_id",
    });
    expect(fds).toBeDefined();
  });

  it("Creation without an id generator function throws", () => {
    expect(
      () =>
        new FileDataSource<string, Device>({
          filePath: "somefile.json",
          idGenerator: "something_that_is_not_a_function" as any as IdGeneratorFunction<string>,
        })
    ).toThrow("The idGenerator must be a function!");
  });
});

describe("FileDataSource CRUD tests", () => {
  let dataSource: FileDataSource<string, Device> = new FileDataSource<string, Device>({
    filePath: "somefile.json",
    idGenerator: () => "the_id",
  });

  beforeEach(() => {
    jest.clearAllMocks();
    jest.resetAllMocks();

    mockExistsSync.mockReturnValue(true);
  });

  it("When called with an non-existing file, the database is created as empty first", async () => {
    mockExistsSync.mockReturnValue(false);
    mockReadFile.mockResolvedValue("[]");

    await dataSource.create({
      id: "",
      series: "the-series",
      modelName: "the-model",
      deviceConfiguration: {
        deviceType: DeviceType.EnvironmentalSensor,
      },
    });

    expect(mockExistsSync).toHaveBeenCalledTimes(1);
    expect(mockWriteFile).toHaveBeenNthCalledWith(1, "[]");
    expect(mockWriteFile).toHaveBeenCalledTimes(2);
    expect(mockReadFile).toHaveBeenCalledTimes(1);
  });

  it("When called with an existing file, the database is read properly", async () => {
    mockReadFile.mockResolvedValue("[]");

    await dataSource.create({
      id: "",
      series: "the-series",
      modelName: "the-model",
      deviceConfiguration: {
        deviceType: DeviceType.EnvironmentalSensor,
      },
    });

    expect(mockExistsSync).toHaveBeenCalledTimes(1);
    expect(mockWriteFile).toHaveBeenCalledTimes(1);
    expect(mockWriteFile.mock.calls[0][0]).toMatchSnapshot();
    expect(mockReadFile).toHaveBeenCalledTimes(1);
  });

  it("Create adds the item to the array", async () => {
    mockReadFile.mockResolvedValue(DEFAULT_DATABASE);

    const device = await dataSource.create({
      id: "",
      series: "the-series",
      modelName: "the-model",
      deviceConfiguration: {
        deviceType: DeviceType.EnvironmentalSensor,
      },
    });

    expect(device).toMatchSnapshot();
    expect(mockWriteFile).toHaveBeenCalledTimes(1);
    expect(mockWriteFile.mock.calls[0][0]).toMatchSnapshot();
  });

  it("Read returns a device successfully", async () => {
    mockReadFile.mockResolvedValue(DEFAULT_DATABASE);

    const device = await dataSource.read("32b3e36c-2e86-4f03-9a09-0b6c676057b3");

    expect(device).toMatchSnapshot();
    expect(mockReadFile).toHaveBeenCalledTimes(1);
    expect(mockWriteFile).not.toHaveBeenCalled();
  });

  it("Read throws an error when the device doesn't exists", async () => {
    mockReadFile.mockResolvedValue(DEFAULT_DATABASE);

    await expect(dataSource.read("some-random-string")).rejects.toThrow(
      "Item with id 'some-random-string' not found!"
    );
    expect(mockReadFile).toHaveBeenCalledTimes(1);
    expect(mockWriteFile).not.toHaveBeenCalled();
  });

  it("List returns the list of all items", async () => {
    mockReadFile.mockResolvedValue(DEFAULT_DATABASE);

    const result = await dataSource.list((_item) => true);

    expect(result).toMatchSnapshot();
    expect(mockReadFile).toHaveBeenCalledTimes(1);
    expect(mockWriteFile).not.toHaveBeenCalled();
  });

  it("List returns the list of all items without supplying a predicate", async () => {
    mockReadFile.mockResolvedValue(DEFAULT_DATABASE);

    const result = await dataSource.list();

    expect(result).toMatchSnapshot();
    expect(mockReadFile).toHaveBeenCalledTimes(1);
    expect(mockWriteFile).not.toHaveBeenCalled();
  });

  it("List returns the list of filtered items", async () => {
    mockReadFile.mockResolvedValue(DEFAULT_DATABASE);

    const result = await dataSource.list(
      (item) => item.id !== "992a28d9-c78d-416c-95df-c65c2ead406c"
    );

    expect(result).toMatchSnapshot();
    expect(mockReadFile).toHaveBeenCalledTimes(1);
    expect(mockWriteFile).not.toHaveBeenCalled();
  });

  it("Update updates a device successfully", async () => {
    mockReadFile.mockResolvedValue(DEFAULT_DATABASE);

    const result = await dataSource.update({
      deviceConfiguration: {
        concentrationHighThreshold: 0.5,
        deviceType: DeviceType.SmokeAlarm,
      },
      id: "32b3e36c-2e86-4f03-9a09-0b6c676057b3",
      modelName: "new_name",
      series: "new_series",
    });

    expect(result).toMatchSnapshot();
    expect(mockReadFile).toHaveBeenCalledTimes(1);
    expect(mockWriteFile).toHaveBeenCalledTimes(1);
    expect(mockWriteFile.mock.calls[0][0]).toMatchSnapshot();
  });

  it("Update throws if a device doesn't exist", async () => {
    mockReadFile.mockResolvedValue(DEFAULT_DATABASE);

    await expect(
      dataSource.update({
        deviceConfiguration: {
          concentrationHighThreshold: 0.5,
          deviceType: DeviceType.SmokeAlarm,
        },
        id: "non-existing-id",
        modelName: "new_name",
        series: "new_series",
      })
    ).rejects.toThrow("Item with id 'non-existing-id' not found!");

    expect(mockReadFile).toHaveBeenCalledTimes(1);
    expect(mockWriteFile).not.toHaveBeenCalled();
  });

  it("Delete deletes a device successfully", async () => {
    mockReadFile.mockResolvedValue(DEFAULT_DATABASE);

    const result = await dataSource.delete("32b3e36c-2e86-4f03-9a09-0b6c676057b3");

    expect(result).toBeTruthy();
    expect(mockReadFile).toHaveBeenCalledTimes(1);
    expect(mockWriteFile).toHaveBeenCalledTimes(1);
    expect(mockWriteFile.mock.calls[0][0]).toMatchSnapshot();
  });

  it("Delete throws an error if device doesn't exist", async () => {
    mockReadFile.mockResolvedValue(DEFAULT_DATABASE);

    await expect(dataSource.delete("non-existing-id2")).rejects.toThrow(
      "Item with id 'non-existing-id2' not found!"
    );

    expect(mockReadFile).toHaveBeenCalledTimes(1);
    expect(mockWriteFile).not.toHaveBeenCalled();
  });
});
