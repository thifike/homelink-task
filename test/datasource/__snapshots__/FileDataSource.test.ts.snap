// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileDataSource CRUD tests Create adds the item to the array 1`] = `
{
  "deviceConfiguration": {
    "deviceType": "ES",
  },
  "id": "the_id",
  "modelName": "the-model",
  "series": "the-series",
}
`;

exports[`FileDataSource CRUD tests Create adds the item to the array 2`] = `"[{"series":"3000","modelName":"Ei3014 Heat Alarm","deviceConfiguration":{"deviceType":"HA","temperatureLowThreshold":16,"temperatureHighThreshold":31},"id":"4796cc2e-1015-469d-87f1-f7a8a7c7a991"},{"series":"1000","modelName":"Ei1020 Environmental Sensor","deviceConfiguration":{"deviceType":"ES"},"id":"992a28d9-c78d-416c-95df-c65c2ead406c"},{"series":"3000","modelName":"Ei3016 Optical Smoke Alarm","deviceConfiguration":{"deviceType":"SA","concentrationHighThreshold":0.4},"id":"32b3e36c-2e86-4f03-9a09-0b6c676057b3"},{"id":"the_id","series":"the-series","modelName":"the-model","deviceConfiguration":{"deviceType":"ES"}}]"`;

exports[`FileDataSource CRUD tests Delete deletes a device successfully 1`] = `"[{"series":"3000","modelName":"Ei3014 Heat Alarm","deviceConfiguration":{"deviceType":"HA","temperatureLowThreshold":16,"temperatureHighThreshold":31},"id":"4796cc2e-1015-469d-87f1-f7a8a7c7a991"},{"series":"1000","modelName":"Ei1020 Environmental Sensor","deviceConfiguration":{"deviceType":"ES"},"id":"992a28d9-c78d-416c-95df-c65c2ead406c"}]"`;

exports[`FileDataSource CRUD tests List returns the list of all items 1`] = `
[
  {
    "deviceConfiguration": {
      "deviceType": "HA",
      "temperatureHighThreshold": 31,
      "temperatureLowThreshold": 16,
    },
    "id": "4796cc2e-1015-469d-87f1-f7a8a7c7a991",
    "modelName": "Ei3014 Heat Alarm",
    "series": "3000",
  },
  {
    "deviceConfiguration": {
      "deviceType": "ES",
    },
    "id": "992a28d9-c78d-416c-95df-c65c2ead406c",
    "modelName": "Ei1020 Environmental Sensor",
    "series": "1000",
  },
  {
    "deviceConfiguration": {
      "concentrationHighThreshold": 0.4,
      "deviceType": "SA",
    },
    "id": "32b3e36c-2e86-4f03-9a09-0b6c676057b3",
    "modelName": "Ei3016 Optical Smoke Alarm",
    "series": "3000",
  },
]
`;

exports[`FileDataSource CRUD tests List returns the list of all items without supplying a predicate 1`] = `
[
  {
    "deviceConfiguration": {
      "deviceType": "HA",
      "temperatureHighThreshold": 31,
      "temperatureLowThreshold": 16,
    },
    "id": "4796cc2e-1015-469d-87f1-f7a8a7c7a991",
    "modelName": "Ei3014 Heat Alarm",
    "series": "3000",
  },
  {
    "deviceConfiguration": {
      "deviceType": "ES",
    },
    "id": "992a28d9-c78d-416c-95df-c65c2ead406c",
    "modelName": "Ei1020 Environmental Sensor",
    "series": "1000",
  },
  {
    "deviceConfiguration": {
      "concentrationHighThreshold": 0.4,
      "deviceType": "SA",
    },
    "id": "32b3e36c-2e86-4f03-9a09-0b6c676057b3",
    "modelName": "Ei3016 Optical Smoke Alarm",
    "series": "3000",
  },
]
`;

exports[`FileDataSource CRUD tests List returns the list of filtered items 1`] = `
[
  {
    "deviceConfiguration": {
      "deviceType": "HA",
      "temperatureHighThreshold": 31,
      "temperatureLowThreshold": 16,
    },
    "id": "4796cc2e-1015-469d-87f1-f7a8a7c7a991",
    "modelName": "Ei3014 Heat Alarm",
    "series": "3000",
  },
  {
    "deviceConfiguration": {
      "concentrationHighThreshold": 0.4,
      "deviceType": "SA",
    },
    "id": "32b3e36c-2e86-4f03-9a09-0b6c676057b3",
    "modelName": "Ei3016 Optical Smoke Alarm",
    "series": "3000",
  },
]
`;

exports[`FileDataSource CRUD tests Read returns a device successfully 1`] = `
{
  "deviceConfiguration": {
    "concentrationHighThreshold": 0.4,
    "deviceType": "SA",
  },
  "id": "32b3e36c-2e86-4f03-9a09-0b6c676057b3",
  "modelName": "Ei3016 Optical Smoke Alarm",
  "series": "3000",
}
`;

exports[`FileDataSource CRUD tests Update updates a device successfully 1`] = `
{
  "deviceConfiguration": {
    "concentrationHighThreshold": 0.5,
    "deviceType": "SA",
  },
  "id": "32b3e36c-2e86-4f03-9a09-0b6c676057b3",
  "modelName": "new_name",
  "series": "new_series",
}
`;

exports[`FileDataSource CRUD tests Update updates a device successfully 2`] = `"[{"series":"3000","modelName":"Ei3014 Heat Alarm","deviceConfiguration":{"deviceType":"HA","temperatureLowThreshold":16,"temperatureHighThreshold":31},"id":"4796cc2e-1015-469d-87f1-f7a8a7c7a991"},{"series":"1000","modelName":"Ei1020 Environmental Sensor","deviceConfiguration":{"deviceType":"ES"},"id":"992a28d9-c78d-416c-95df-c65c2ead406c"},{"deviceConfiguration":{"concentrationHighThreshold":0.5,"deviceType":"SA"},"id":"32b3e36c-2e86-4f03-9a09-0b6c676057b3","modelName":"new_name","series":"new_series"}]"`;

exports[`FileDataSource CRUD tests When called with an existing file, the database is read properly 1`] = `"[{"id":"the_id","series":"the-series","modelName":"the-model","deviceConfiguration":{"deviceType":"ES"}}]"`;
